### DISCUSSION TITLE
The Future of Schedulers in Mesa

### DISCUSSION THREAD

**Discussion by EwoutH** [mesa/mesa#2231] [Future of Mesa]
_**Update:** View the [migration guide](https://mesa.readthedocs.io/en/latest/migration_guide.html#time-and-schedulers) on how to replace the deprecated Schedulers with AgentSet functionality._

_This discussion is split of from_
- https://github.com/projectmesa/mesa/pull/2223

_This discussion builds on:_
- https://github.com/projectmesa/mesa/discussions/1912

## Background
Mesa currently uses various scheduler classes to manage agent activation. There's ongoing discussion about the role and implementation of schedulers in future versions.

## Key Points to Consider

1. **Current Scheduler Types**: Evaluating the usefulness and limitations of existing schedulers (e.g., RandomActivation, SimultaneousActivation).
2. **Flexibility vs. Simplicity**: Balancing the need for flexible scheduling with keeping Mesa simple for beginners.
3. **Integration with Time Systems**: How schedulers should interact with step counters or potential new time/clock systems.
4. **Alternative Approaches**: Exploring other paradigms, such as event-driven simulations or agent-set based approaches.

### Mesa 2 approach:
```python
class MyModel(Model):
    def __init__(self):
        self.schedule = RandomActivation(self)
        # Add agents to schedule

    def step(self):
        self.schedule.step()
```

### Mesa 3 approach:
```python
class MyModel(Model):
    def __init__(self):
        ...

    def step(self):
        self.agents.shuffle().do("step")
```

## Questions for Discussion

1. Are the current scheduler classes meeting the needs of Mesa users? If not, what's missing?
2. Should schedulers be a core part of Mesa, or should they be optional/pluggable components?
3. How can we design a scheduling system that's simple for beginners but flexible for advanced users?
4. Should Mesa move towards a more event-driven or agent-set based approach instead of the current scheduler system?
5. How can we ensure that any changes to the scheduler system don't break existing models or make them significantly more complex?

**Comment by EwoutH** [comment]
I think my stance in this is clear: I don't see value in schedulers in Mesa now that we have the AgentSet and an internal clock in Mesa. They can be beautifully replaced, and as an example of that, I want to let this PR speak for itself:

- https://github.com/projectmesa/mesa-examples/pull/161

We write a clear migration guide how to replace the schedulers with AgentSet operations. If you want to keep using schedulers, that's fine, Mesa 2.x is still there.

**Reply by quaquel** [reply]
I am coming around to this point of view as well. With AgentSet, the old schedulers are redundant. What remains is agreeing on the handling of the simulation clock.

**Reply by rht** [reply]
To speed up the consensus process, yes, I'm up for the removal of the schedulers from time.py today. It's just that the performance regression of the `model.agents` must be fixed before the release.

**Reply by quaquel** [reply]
I'll try to get #2220 finished first based on the useful feedback received there and then I'll turn my attention to this. I have a few possible solutions I wan't to benchmark and test for their API.

**Reply by EwoutH** [reply]
I'm going to open a PR soon to to deprecate most schedulers. The AgentSet offers all functionality, and thus our holy book ([PEP 20](https://peps.python.org/pep-0020/)) can be followed:

> There should be one-- and preferably only one --obvious way to do it.

And a bit of 

> Explicit is better than implicit.

And, personally:

> Beautiful is better than ugly.
> Complex is better than complicated.
> Readability counts.
